plugin/src/main/java/org/opensearch/ml/utils/SecurityUtils.java

```diff
@@ -0,0 +1,122 @@
+                    }
+                    if (source.containsKey(MLModelGroup.OWNER)) {
+                        owner = (Map) source.get(MLModelGroup.OWNER);
+                    }
+                    if (access != null && StringUtils.equals(access, MLModelGroup.PUBLIC)) {
+                        listener.onResponse(Boolean.TRUE);
+                    } else if (access != null && StringUtils.equals(access, MLModelGroup.PRIVATE)) {
+                        if (isOwner(owner, user)) {
+                            listener.onResponse(Boolean.TRUE);
+                            return;
+                        }
+                        listener.onResponse(Boolean.FALSE);
+                    } else {
+                        listener
+                            .onResponse(
+                                userBackendRoles.stream().anyMatch(modelGroupBackendRoles.stream().collect(Collectors.toSet())::contains)
+                            );
+                    }
+                }
+            }, e -> {
+                log.error("Failed to validate Access", e);
+                listener.onFailure(new MLValidationException("Failed to validate Access"));
+            }));
+        } catch (Exception e) {
+            log.error("Failed to validate Access", e);
+            listener.onFailure(e);
+        }
+    }
+
+    public static boolean isAdmin(User user) {
+        if (user == null) {
+            return false;
+        }
+        if (CollectionUtils.isEmpty(user.getRoles())) {
+            return false;
+        }
+        return user.getRoles().contains("all_access");
+    }
+
+    public static boolean isOwner(Map<String, Object> owner, User user) {
+        if (user == null || owner == null || !owner.containsKey("name")) {
+            return false;
+        }
+        return StringUtils.equals((String) owner.get("name"), user.getName());
+    }
+
+    public static SearchSourceBuilder addUserBackendRolesFilter(User user, SearchSourceBuilder searchSourceBuilder) {
+        if (user == null) {
+            return searchSourceBuilder;
+        }
+        BoolQueryBuilder boolQueryBuilder = new BoolQueryBuilder();
+        String ownerBackendRoleFieldName = "backend_roles.keyword";
+        List<String> backendRoles = user.getBackendRoles() != null ? user.getBackendRoles() : ImmutableList.of();
+        TermsQueryBuilder userRolesFilterQuery = QueryBuilders.termsQuery(ownerBackendRoleFieldName, backendRoles);
+        boolQueryBuilder.must(userRolesFilterQuery);
+        QueryBuilder query = searchSourceBuilder.query();
+        if (query == null) {
+            searchSourceBuilder.query(boolQueryBuilder);
+        } else if (query instanceof BoolQueryBuilder) {
+            ((BoolQueryBuilder) query).filter(boolQueryBuilder);
+        } else {
+            throw new MLValidationException("Search API does not support queries other than BoolQuery");
```

Comment:
We do not support match/match_phrase/term query in search API? This doesn't seem correct.

Reply:
This could cause issue since frontend will be using these queries, we need more discussion on this to make sure there's no blocker for other requirements.

Reply:
@ylwu any inputs here?